core/unix/os.c

```diff
@@ -7444,26 +7445,37 @@ pre_system_call(dcontext_t *dcontext)
             size_t sizemask;
         } data_t;
         data_t *data = (data_t *)sys_param(dcontext, 5);
-        kernel_sigset_t *mask = (kernel_sigset_t *)data->sigmask;
-        size_t sizemask = (size_t)data->sizemask;
-        kernel_sigset_t nullsigmask;
-        memset(&nullsigmask, 0, sizeof(nullsigmask));
-        /* syscall ABI, struct is in the  app's stack */
-        safe_write_ex((void *)&data->sigmask, sizeof(kernel_sigset_t),
-                      (const void *)&nullsigmask, NULL);
+        kernel_sigset_t *mask;
+        if (!safe_read(&data->sigmask, sizeof(kernel_sigset_t), &mask)) {
+            LOG(THREAD, LOG_SYSCALLS, 2, "\treturning EFAULT to app for pselect6\n");
+            set_failure_return_val(dcontext, EINVAL);
+            DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+        }
+        size_t sizemask;
+        if (!safe_read(&data->sizemask, sizeof(size_t), &sizemask)) {
+            LOG(THREAD, LOG_SYSCALLS, 2, "\treturning EFAULT to app for pselect6\n");
+            set_failure_return_val(dcontext, EINVAL);
+            DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+        }
+        /* We're using sys_param4 to save the sizemask until post syscall to be
+         * able to restore it.
+         */
+        dcontext->sys_param4 = (reg_t)data->sizemask;
+        /* See syscall ABI, struct is in the  app's stack. */
+        safe_write_ex((void *)&data->sigmask, sizeof(kernel_sigset_t), (const void *)NULL,
```

Comment:
another bug: safe_write_ex takes in a *pointer* to a buffer containing values to write.  You're passing NULL as that pointer, so it will crash: but since it's "safe" it will handle the crash and return failure.  You're not checking the return value here, so the failure is not noticed.

Reply:
That's why I previously had a nullsigmask buffer remember? Then I got confused that you wanted to change that.

Reply:
No, there should not be any nullsigmask.  This is a pointer, not a struct.  The issue is that safe_write_ex, just like memcpy, takes in a pointer to the value (otherwise how could it handle arbitrary sizes if it took the value to write by value?).  You have to declare a kernel_sigset_t* and set it to NULL and point at its address.

Reply:
Done